### ISSUE TITLE
Ensure that we have timeouts in place when reading/writing nodes in IPFS

### ISSUE THREAD

**Issue by aphelionz** [orbitdb/ipfs-log#199]
From: https://github.com/ipfs/dynamic-data-and-capabilities/issues/50

Ensure that we have timeouts in place when reading/writing nodes in IPFS

**Comment by aphelionz** [comment]
cc @satazor

**Comment by satazor** [comment]
@aphelionz Should I take care of this one or wait for #213? I'm asking because there will be some conflicts.

**Comment by aphelionz** [comment]
Sure, go ahead

**Comment by satazor** [comment]
Before I actually implement this, I will need some feedback on the best way to implement add this feature. I will use the [`p-timeout`](https://github.com/sindresorhus/p-timeout) module but consumers of the module need to be in control of the actual timeout in ms. The default should be `no timeout` to keep things unchanged but what's the best way to pass the actual timeout?

A Log can be instantiated directly via its constructor or via static functions, such as `fromMultihash`. This means that the static functions already must include the ability to specify the read timeout because they perform I/O. The most practical way is to add a `timeout` argument to all of these functions but I'm worried about the number of arguments. As an example:

```js
static async fromMultihash (ipfs, access, identity, hash, length = -1, exclude, onProgressCallback) {
```

already has 7 arguments, 3 of them being optional. I think that the most "javascripty" way to do this was:

```js
static async fromMultihash (ipfs, access, identity, hash, options) {
```

where `options` would look like:

```
{
  length,
  exclude,
  onProgress,
}
```

This way, we could add more optional arguments, like the `timeout` argument I'm proposing.

-------------------

I would love to hear more about this and how to proceed.

**Comment by aphelionz** [comment]
@satazor Let's keep it as a discrete argument for now, even though there's a lot.

We want to favor a transparent function signature that doesn't "hide" the arguments in an `options` object. Even with the new syntax that exposes the object members, it it's still hard to tell what options are available. At least with the large number of discrete arguments, it's easy for the user to tell what they are, and what they expect.

The reminds me, I need to add the `sortFn` argument to those functions too!

**Comment by satazor** [comment]
I understand your concerns but currently it’s very cumbursome to specify a value to an optional argument specially if it’s the last one: one has specify undefined for all the arguments that want to keep the default.

Of course that all this topic is subjective and boils down to personal preference, but my personal view is that in a language that doesn’t support method overloading, something like my proposal is better than what we have now.

Nevertheless, I will advance with adding the argument to the end of the list.

**Comment by haadcode** [comment]
Not sure if does exactly what you need @satazor, but we do have per-entry-load/fetch timeout in https://github.com/orbitdb/ipfs-log/blob/master/src/entry-io.js#L65.

**Comment by aphelionz** [comment]
This is the best example I could find for the destructured function parameters, essentially using the new destructured assignment notation to simulate named parameters. I think we may be able to achieve the best of both worlds with this approach. http://2ality.com/2015/01/es6-destructuring.html#simulating-named-parameters-in-javascript

**Comment by satazor** [comment]
@aphelionz with destructuring, the "arguments" are still pretty visible. Neverthless, I shall proceed with adding another argument in the PR and later we discuss refactor the optional arguments into `options` right?

**Comment by satazor** [comment]
As @haadcode pointed out, we already have a `timeout` argument in `Entry.fetchAll` & `Entry.fetchParallel`. In these functions, the `timeout` argument comes before the `onProgressCallback` and it should be like this everywhere to maintain consistency. By everywhere, I mean: `LogIO.fromEntryCid`, `LogIO.fromJSON`, `LogIO.fromEntry`, `Log.fromEntryCid`, `Log.fromJSON`, `Log.fromEntry`. The problem is that if I add before, it will be a breaking change for the public ones. Solutions:

- Break consistency and put `timeout` after the `onProgressCallback` on `LogIO.fromEntryCid`, `LogIO.fromJSON`, `LogIO.fromEntry`, `Log.fromEntryCid`, `Log.fromJSON`, `Log.fromEntry`
- Move `timeout` after `onProgressCallback` on `Entry.fetchAll` & `Entry.fetchParallel`; we can do this because it's not part of the public API
- Place `timeout` before `onProgressCallback` but do a typeof check in case `timeout` is a function, for backwards compatibility
- Adopt the `options` strategy I mentioned above in a backwards compatible manner, which doesn't suffer from the positioning issue I'm detailing here; We could create a util/helper that would ensure the backwards compatibility and would be reused in a lot of places.

Let me know what your thoughts are @aphelionz @haadcode @shamb0t

**Comment by aphelionz** [comment]
@satazor @shamb0t sure, I can address this in my work in https://github.com/orbitdb/ipfs-log/issues/219. Go ahead and put `timeout` after `onProgressCallback` to get this resolved.

**Comment by satazor** [comment]
PR at #219 but I don't really like the looks of it because the `Log.fromJSON` already exposed the `timeout` before `onProgressCallback`. This means that the public API is inconsistent. Should we wait for #219 that brings the options?

**Comment by aphelionz** [comment]
That makes sense to me @satazor

**Comment by aphelionz** [comment]
@satazor My PR #220 has been merged. You should be able to work your timeouts into the destructure options object now.

**Comment by aphelionz** [comment]
Removing my assignment as I'm not currently working on this.

**Comment by shamb0t** [comment]
Resolved by #258
